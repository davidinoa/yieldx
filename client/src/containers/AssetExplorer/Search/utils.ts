interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
