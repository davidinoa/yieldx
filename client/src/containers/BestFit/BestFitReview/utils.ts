import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
