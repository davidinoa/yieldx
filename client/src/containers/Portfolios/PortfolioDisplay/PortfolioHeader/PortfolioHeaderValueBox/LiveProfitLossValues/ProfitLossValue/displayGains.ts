import Decimal from "../../../../../../../util/Decimal";
import { roundToHundredth } from "../../../../../../../util/roundToHundredth";

export function displayGains(gain: number) {
  // rounding to the nearest 10th place
  const roundedGain = roundToHundredth(gain);
  if (roundedGain === 0) return "$0.00";
  let displayValue = roundedGain > 0 ? "+ $" : "- $";
  displayValue += Decimal.format(Math.abs(roundedGain));
  return displayValue;
}
