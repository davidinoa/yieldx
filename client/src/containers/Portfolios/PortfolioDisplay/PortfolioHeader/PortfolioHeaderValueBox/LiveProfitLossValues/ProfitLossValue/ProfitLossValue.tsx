import React from "react";
import clsx from "clsx";
import { Spinner } from "elements/Spinner/Spinner";
import { ReactComponent as RedAlert } from "assets/svg/red-alert.svg";
import { displayGains } from "./displayGains";

interface Props {
  label: string;
  value?: number;
  error?: boolean;
}

export default function ProfitLossValue({ value, error, label }: Props) {
  return (
    <div className="d-inline-block performance">
      <span className="label">{label}</span>
      {value !== undefined && !error && (
        <span
          className={clsx("value", {
            positive: value > 0,
            negative: value < 0
          })}
        >
          {displayGains(value)}
        </span>
      )}
      {value === undefined && !error && (
        <span>
          <Spinner />
        </span>
      )}
      {error && (
        <span>
          <RedAlert />
        </span>
      )}
    </div>
  );
}
