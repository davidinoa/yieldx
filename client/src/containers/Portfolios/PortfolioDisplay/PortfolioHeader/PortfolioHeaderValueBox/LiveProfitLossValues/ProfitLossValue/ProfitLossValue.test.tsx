import React from "react";
import { render, screen } from "../../../../../../../test/utils";
import ProfitLossValue from "./ProfitLossValue";

describe("ProfitLossValue", () => {
  test("Should render label correctly", () => {
    render(<ProfitLossValue label="today gain/loss:" />);
    expect(screen.queryByText(/today gain\/loss:/i)).toBeInTheDocument();
  });

  test("Should render spinner if value undefined", () => {
    render(<ProfitLossValue label="today gain/loss:" />);
    expect(screen.queryByTestId("spinner")).toBeInTheDocument();
  });

  test("Should render formatted value prop with correct class", () => {
    const { rerender, asFragment } = render(
      <ProfitLossValue label="today gain/loss:" value={3.5} />
    );
    expect(asFragment()).toMatchSnapshot();
    rerender(<ProfitLossValue label="today gain/loss:" value={0} />);
    expect(asFragment()).toMatchSnapshot();
    rerender(<ProfitLossValue label="today gain/loss:" value={-3.5} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("Should render error icon if error prop is true", () => {
    const { asFragment } = render(
      <ProfitLossValue label="today gain/loss:" error />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
