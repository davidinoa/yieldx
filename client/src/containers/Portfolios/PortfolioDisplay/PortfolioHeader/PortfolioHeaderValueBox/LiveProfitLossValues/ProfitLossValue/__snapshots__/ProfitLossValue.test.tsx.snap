// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfitLossValue Should render error icon if error prop is true 1`] = `
<DocumentFragment>
  <div
    class="d-inline-block performance"
  >
    <span
      class="label"
    >
      today gain/loss:
    </span>
    <span>
      <svg>
        red-alert.svg
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`ProfitLossValue Should render formatted value prop with correct class 1`] = `
<DocumentFragment>
  <div
    class="d-inline-block performance"
  >
    <span
      class="label"
    >
      today gain/loss:
    </span>
    <span
      class="value positive"
    >
      + $3.50
    </span>
  </div>
</DocumentFragment>
`;

exports[`ProfitLossValue Should render formatted value prop with correct class 2`] = `
<DocumentFragment>
  <div
    class="d-inline-block performance"
  >
    <span
      class="label"
    >
      today gain/loss:
    </span>
    <span
      class="value"
    >
      $0.00
    </span>
  </div>
</DocumentFragment>
`;

exports[`ProfitLossValue Should render formatted value prop with correct class 3`] = `
<DocumentFragment>
  <div
    class="d-inline-block performance"
  >
    <span
      class="label"
    >
      today gain/loss:
    </span>
    <span
      class="value negative"
    >
      - $3.50
    </span>
  </div>
</DocumentFragment>
`;
